src/lib/transaction_snark/test/account_timing/account_timing.ml

```diff
@@ -837,4 +813,601 @@ let%test_module "account timing check" =
+              (* high init min balance, payment amount enough to violate *)
+              let (timing : Account_timing.t) =
+                Timed
+                  { initial_minimum_balance =
+                      Currency.Balance.of_int 100_000_000_000_000
+                  ; cliff_time = Mina_numbers.Global_slot.of_int 10000
+                  ; cliff_amount = Currency.Amount.of_int 100
+                  ; vesting_period = Mina_numbers.Global_slot.of_int 1
+                  ; vesting_increment = Currency.Amount.of_int 10
+                  }
+              in
+              let balance_as_amount = Currency.Balance.to_amount balance in
+              (keypair, balance_as_amount, nonce, timing))
+          |> Array.of_list
+        in
+        let parties_command =
+          let open Mina_base in
+          let fee = Currency.Fee.of_int 1_000_000 in
+          let amount = Currency.Amount.of_int 10_000_000_000_000 in
+          let nonce = Account.Nonce.zero in
+          let memo =
+            Signed_command_memo.create_from_string_exn
+              "Snapp transfer, timed account"
+          in
+          let sender_keypair = List.hd_exn keypairs in
+          let zkapp_keypair = List.nth_exn keypairs 1 in
+          let receiver_key =
+            zkapp_keypair.public_key |> Signature_lib.Public_key.compress
+          in
+          let (parties_spec : Transaction_snark.For_tests.Spec.t) =
+            { sender = (sender_keypair, nonce)
+            ; fee
+            ; receivers = [ (receiver_key, amount) ]
+            ; amount
+            ; snapp_account_keypairs = []
+            ; memo
+            ; new_snapp_account = false
+            ; snapp_update = Party.Update.dummy
+            ; current_auth = Permissions.Auth_required.Signature
+            ; call_data = Snark_params.Tick.Field.zero
+            ; events = []
+            ; sequence_events = []
+            }
+          in
+          Transaction_snark.For_tests.multiple_transfers parties_spec
+        in
+        return (ledger_init_state, parties_command)
+      in
+      (* slot 1, well before cliffs *)
+      Quickcheck.test ~seed:(`Deterministic "zkapp command, before cliff")
+        ~sexp_of:[%sexp_of: Mina_ledger.Ledger.init_state * Parties.t] ~trials:1
+        gen ~f:(fun (ledger_init_state, parties) ->
+          Mina_ledger.Ledger.with_ephemeral_ledger
+            ~depth:constraint_constants.ledger_depth ~f:(fun ledger ->
+              Mina_ledger.Ledger.apply_initial_ledger_state ledger
+                ledger_init_state ;
+              let _state_body, state_view =
+                state_body_and_view_at_slot Mina_numbers.Global_slot.(succ zero)
+              in
+              match
+                Mina_ledger.Ledger.apply_parties_unchecked ~constraint_constants
+                  ~state_view ledger parties
+              with
+              | Ok _txn_applied ->
+                  failwith "Should have failed with min balance violation"
```

Comment:
Can you add a case where the transaction succeeds but returns `Failed` status? The failure_status_tbl should have the min balance violation error. In the test_spec you could add a separate field for fee_payer and make `sender` an option. If sender is `None` then sender=fee_payer. That way you create a txns that is accepted but the party update fails.

Reply:
I mean `Ok (Failed ...)` instead of `Error ...`. The transaction fails only if fee payer violates min the min balance. You could add another case where the fee payer account has enough balance to pay the fee but not make any other transfers

Reply:
I think that's what happens for the two tests that call `check_zkapp_failure`: the `apply` returns `Ok`, but the transaction status inside the `Ok` is `Failed`.

Reply:
Those are for `while vesting` and `after vesting`.  For completeness sake, should we be add one for `before cliff`?

Reply:
I modified the existing min balance violation test for `before cliff` to call `check_zkapp_failure`.

I had to change the initial min balance to something less than the account balance. If it was the same as the account balance, I got an `Error` instead of `Ok`, and the error string contained a tag.